main.ts

```diff
@@ -1,150 +1,293 @@
+      [`${delimStr}+`, delimStr],
+      [`${delimStr}(?!\ )`, titleSeparator],
+      [`(?<!\ )${delimStr}`, ''],
+    ]);
+    for (const [key, value] of replacements) {
+      let re = new RegExp(key, 'g');
+      title = title.replace(re, value);
+    }
+    return title
+  };
+
+  private readonly handleRename = async (file: TFile, oldPath: string): Promise<void> => {
+    // console.log(`file: ${oldPath} renamed to: ${file.path}`);
+    if (file instanceof TFile && file === this.app.workspace.getActiveFile()) {
+      this.app.metadataCache.onCleanCache(() => { this.refreshTitle(file); });
+    }
+  };
+
+  private readonly handleDelete = async (file: TFile): Promise<void> => {
+    this.refreshTitle();
+  };
+
+  private readonly handleOpen = async (file: TFile): Promise<void> => {
+    if (file instanceof TFile && file === this.app.workspace.getActiveFile()) {
+      this.refreshTitle(file);
+    }
+  };
+
+  private readonly handleMetaChange = async (file: TFile): Promise<void> => {
+    if (file instanceof TFile && file === this.app.workspace.getActiveFile()) {
+      this.refreshTitle(file);
+    }
+  };
+
+  private readonly handleMetaResolve = async (file: TFile): Promise<void> => {
+    if (file instanceof TFile && file === this.app.workspace.getActiveFile()) {
+      this.refreshTitle(file);
+    }
+  };
+
+  async loadSettings() {
+    this.settings = Object.assign({}, DEFAULT_SETTINGS, await this.loadData());
+  }
+
+  async saveSettings() {
+    await this.saveData(this.settings);
+  }
+
+}
+
+interface ActiveNoteTitlePluginSettings {
+  titleTemplate: string,
+  titleTemplateEmpty: string,
+  titleSeparator: string,
+  delimStr: string
+}
+
+const DEFAULT_SETTINGS: ActiveNoteTitlePluginSettings = {
+  titleTemplate: "{{basename}}~~{{vault}} - Obsidian v{{version}}",
+  titleTemplateEmpty: "{{vault}} - Obsidian v{{version}}",
+  titleSeparator: " - ",
+  delimStr: "~~"
+}
+
+class ActiveNoteTitlePluginSettingsTab extends PluginSettingTab {
+
+  plugin: ActiveNoteTitlePlugin;
+
+  constructor(app: App, plugin: ActiveNoteTitlePlugin) {
+    super(app, plugin);
+    this.plugin = plugin;
+  }
+
+  display(): void {
+    let { containerEl } = this;
+    let desc: DocumentFragment;
+    containerEl.empty();
+    containerEl.createEl('h2', {text: 'Window title templates'});
+    containerEl.createEl('p', {text: 'These two templates override the window title of the Obsidian window. This is useful for example when you use tracking software that works with window titles. You can use the format `~~{{placeholder}}~~` if you want the placeholder to be completely omitted when blank, otherwise whitespace and other characters will be preserved. You can surround a placeholder with parentheses e.g. `({{frontmatter.project}})` and it will be hidden if the referenced field is empty.'});
+
+    desc = document.createDocumentFragment();
+    desc.append('Available ');
+    desc.createEl('b').innerText = 'placeholders:';
+    let placeholders = [
```

Comment:
These placeholders are now joined on each line, resulting in just each line being wrapped in `{{ }}` instead of each key. A nested forEach might do the job!

Reply:
That was intentional, just trying to save some space in the settings window, I grouped the placeholders together by the "related" ones e.g. all the filename/extension variants on one line. Originally I had them listed out separately but as we added more, the list got very tall. Is this what you're referring to?
![image](https://user-images.githubusercontent.com/1992842/129919581-3940b815-7bec-4e1e-b1d8-96466f513787.png)

Reply:
Yeah, I think it might be confusing... If we change it to a nested loop it might result in:

```
{{vault}},{{workspace}},{{version}}
{{filename}},{{filepath}},{{basename}},{{extension}}
{{frontmatter.<any_frontmatter_key>}}
```

This makes is more clear to the user I think :)